src/v/storage/fs_finject.cc

```diff
@@ -0,0 +1,108 @@
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "fs_finject.h"
+
+#include "seastarx.h"
+#include "storage/logger.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/io_priority_class.hh>
+namespace storage {
+
+thread_local fs_failure_probe fs_finject::_probe = {}; // NOLINT
+using methods = fs_failure_probe::methods;
+
+ss::future<size_t> fs_finject::write_dma(
+  uint64_t pos,
+  const void* buffer,
+  size_t len,
+  const ss::io_priority_class& pc) {
+    return _probe.maybe_inject_failure(methods::write_dma, "write_dma")
+      .then([this, pos, buffer, len, pc]() {
+          return get_file_impl(_file)->write_dma(pos, buffer, len, pc);
+      });
+}
+
+ss::future<size_t> fs_finject::write_dma(
+  uint64_t pos, std::vector<iovec> iov, const ss::io_priority_class& pc) {
+    return _probe.maybe_inject_failure(methods::write_dma, "write_dma")
+      .then([this, pos, iov, pc]() {
+          return get_file_impl(_file)->write_dma(pos, iov, pc);
+      });
+}
+
+ss::future<size_t> fs_finject::read_dma(
+  uint64_t pos, void* buffer, size_t len, const ss::io_priority_class& pc) {
+    return get_file_impl(_file)->read_dma(pos, buffer, len, pc);
+}
+
+ss::future<size_t> fs_finject::read_dma(
+  uint64_t pos, std::vector<iovec> iov, const ss::io_priority_class& pc) {
+    return get_file_impl(_file)->read_dma(pos, iov, pc);
+}
+
+ss::future<> fs_finject::flush() { return get_file_impl(_file)->flush(); }
+
+ss::future<struct stat> fs_finject::stat() {
+    return get_file_impl(_file)->stat();
+}
+
+ss::future<> fs_finject::truncate(uint64_t length) {
+    return get_file_impl(_file)->truncate(length);
+}
+
+ss::future<> fs_finject::discard(uint64_t offset, uint64_t length) {
+    return get_file_impl(_file)->discard(offset, length);
+}
+
+ss::future<int> fs_finject::ioctl(uint64_t cmd, void* argp) noexcept {
+    return get_file_impl(_file)->ioctl(cmd, argp);
+}
+
+ss::future<int> fs_finject::ioctl_short(uint64_t cmd, void* argp) noexcept {
+    return get_file_impl(_file)->ioctl_short(cmd, argp);
+}
+
+ss::future<int> fs_finject::fcntl(int op, uintptr_t arg) noexcept {
+    return get_file_impl(_file)->fcntl(op, arg);
+}
+
+ss::future<int> fs_finject::fcntl_short(int op, uintptr_t arg) noexcept {
+    return get_file_impl(_file)->fcntl_short(op, arg);
+}
+
+ss::future<> fs_finject::allocate(uint64_t pos, uint64_t len) {
+    vlog(stlog.info, "fs_finject hooked allocate()");
+    return _probe.maybe_inject_failure(methods::allocate, "allocate")
+      .then([this, pos, len]() {
+          return get_file_impl(_file)->allocate(pos, len);
+      });
+}
+
+ss::future<uint64_t> fs_finject::size() { return get_file_impl(_file)->size(); }
+
+ss::future<> fs_finject::close() { return get_file_impl(_file)->close(); }
+
+std::unique_ptr<ss::file_handle_impl> fs_finject::dup() {
+    return get_file_impl(_file)->dup();
```

Comment:
perhaps it makes sense to build a `default_file_impl` that calls all of the non-override versions of these and then fs_finject will only override a couple.

Reply:
I can take a stab.. maybe as followup patch? Related to @senior7515 idea to create a proper `storage::file` RP interface as a common point for instrumentation / FI, etc.